### PR TITLE
Update confirmation strategy for Almond

### PR DESCRIPTION
**Pull request by sileix**
```
                                                                       yes     ++----------+                                                                                                                              
                                                              +----------------| slot fill||-----------------+                                                                                                            
                                                              |                ++----------+                 |                                                                                                            
                                                              |                                              |                                                                                                            
                                           query       +------------+  no                               +-------++                                                                                                        
                             +-------------------------| has slots? |-----------------------------------||result||                                                                                                        
                             |                         +------------+                                   +-------++                                                                                                        
                             |                                         yes     ++----------+                                                                                                                              
                             |                                +----------------| slot fill||-----------------+                                                                                                            
                             |                                |                ++----------+                 |                                                                                                            
                             |                                |                                              |                                                                                                            
           yes       +---------------+      safe       +------------+  no                               +-------++              +--------+                                                                                
       +-------------| program type? |----------------+| has slots? |-----------------------------------|| echo ||--------------|| exec ||                                                                                
       |             +---------------+     action      +------------+                                   +-------++              +--------+                                                                                
       |                     |                                         yes     ++----------+           +-----------+            +--------+            +--------+                                                          
       |                     |                                +----------------| slot fill||-----------|| confirm ||------------|| exec ||------------|| done ||                                                          
       |                     |                                |                ++----------+           +-----------+            +--------+            +--------+                               -                          
       |                     |                                |                                                                                                                                                           
       |                     |                         +------------+  no                                +-------++             +--------+                                                                                
       |                     +-------------------------| has slots? |------------------------------------|| echo ||-------------|| exec ||                                                                                
       |                                   others      +------------+                                    +-------++             +--------+                                                                                
       |                                                                                                                                                                                                                  
+------------+                                                                                          yes     ++----------+                                                                                             
| confident? |                                                                                 +----------------| slot fill||-----------------+                                                                           
+------------+                                                                                 |                ++----------+                 |                                                                           
       |                                                                                       |                                              |                                                                           
       |                                                                    query       +------------+  no                               +-------++                                                                       
       |                                                      +-------------------------| has slots? |-----------------------------------||result||                                                                       
       |                                                      |                         +------------+                                   +-------++                                                                       
       |                                                      |                                         yes     ++----------+                                                                                             
       |                                                      |                                +----------------| slot fill||-----------------+                       -                                                   
       |                                                      |                                |                ++----------+                 |                                                                           
       |                                                      |                                |                                              |                                                                           
       |               +-----------+                  +---------------+      safe       +------|-----+  no                               +-------++            +--------+                                                 
       +---------------|| confirm ||------------------| program type? |-----------------| has slots? |-----------------------------------|| echo ||------------|| exec ||                                                 
           no          +-----------+                  +---------------+     action      +------------+                                   +-------++            +--------+                                                 
                                                              |                                         yes     ++----------+           +-----------+          +--------+            +--------+                           
                                                              |                                +----------------| slot fill||-----------|| confirm ||----------|| exec ||------------|| done ||                           
                                                              |                                |                ++----------+           +-----------+          +--------+            +--------+                           
                                                              |                                |                                                                                                                          
                                                              |                         +------------+  no        +--------+             +--------+                                                                       
                                                              +-------------------------| has slots? |------------|| exec ||-------------|| done ||                                                                       
                                                                            others      +------------+            +--------+             +--------+                        
```

### CODE REVIEW

**Comment by gcampax** [comment]
Are we applying this to permission rules too? Why?

File: lib/dialogs/rule.js [hunk 1]
```diff
@@ -96,34 +99,35 @@ async function completeProgram(dlg, program, skipConfirmation) {
...
             continue;
         let newIcon = Helpers.getIcon(prim);
         if (newIcon)
             icon = newIcon;
     }
+    dlg.icon = icon;
+
+    let hasSlots = false;
+    for (let [schema, slot] of program.iterateSlots()) {
+        if (slot instanceof ThingTalk.Ast.InputParam ) {
```
Comment (gcampax) [inline lib/dialogs/rule.js]: Does it matter if the slot is already filled here? What about `$?` in filters?

EDIT: sorry, I saw the fix to check `isUndefined` later. The question about filters remains though.

**Comment by sileix** [comment]
> Are we applying this to permission rules too? Why?

Why not? Why is it different from setup and rule?

File: lib/dialogs/rule.js (hunk 1 above)
Comment (sileix) [reply lib/dialogs/rule.js]: `iterateSlots` returns all the slots in the program including the ones in filters, right?

**Comment by gcampax** [comment]
What's missing here?

**Pull request merged**